main/src/main/scala/sbt/internal/IncrementalTest.scala

```diff
@@ -0,0 +1,218 @@
+end TestStatusReporter
+
+private[sbt] object TestStatus:
+  import java.util.Properties
+  def read(f: File): concurrent.Map[String, Digest] =
+    import scala.jdk.CollectionConverters.*
+    val props = Properties()
+    IO.load(props, f)
+    val result = ConcurrentHashMap[String, Digest]()
+    props.asScala.iterator.foreach { case (k, v) => result.put(k, Digest(v)) }
+    result.asScala
+
+  def write(map: collection.Map[String, Digest], label: String, f: File): Unit =
+    IO.writeLines(
+      f,
+      s"# $label" ::
+        map.toList.sortBy(_._1).map { case (k, v) =>
+          s"$k=$v"
+        }
+    )
+end TestStatus
+
+/**
+ * ClassStamper provides `transitiveStamp` method to calculate a unique
+ * fingerprint, which will be used for runtime invalidation.
+ */
+class ClassStamper(
+    classpath: Seq[Attributed[HashedVirtualFileRef]],
+    converter: FileConverter,
+):
+  private val stamps = mutable.Map.empty[String, SortedSet[Digest]]
+  private val vfStamps = mutable.Map.empty[VirtualFileRef, Digest]
+  private lazy val analyses = classpath
+    .flatMap(a => BuildDef.extractAnalysis(a.metadata, converter))
+    .collect { case analysis: Analysis => analysis }
+
+  /**
+   * Given a classpath and a class name, this tries to create a SHA-256 digest.
+   * @param className className to stamp
+   * @param extraHashes additional information to include into the returning digest
+   */
+  private[sbt] def transitiveStamp(className: String, extaHashes: Seq[Digest]): Option[Digest] =
+    val digests = SortedSet(analyses.flatMap(internalStamp(className, _, Set.empty)): _*)
+    if digests.nonEmpty then Some(Digest.sha256Hash(digests.toSeq ++ extaHashes: _*))
+    else None
+
+  private def internalStamp(
+      className: String,
+      analysis: Analysis,
+      alreadySeen: Set[String],
+  ): SortedSet[Digest] =
+    if alreadySeen.contains(className) then SortedSet.empty
+    else
+      stamps.get(className) match
+        case Some(xs) => xs
+        case _ =>
+          import analysis.relations
+          val internalDeps = relations
+            .internalClassDeps(className)
+            .flatMap: otherCN =>
+              internalStamp(otherCN, analysis, alreadySeen + className)
+          val internalJarDeps = relations
+            .externalDeps(className)
+            .flatMap: libClassName =>
+              transitiveStamp(libClassName, Nil)
+          val externalDeps = relations
+            .externalDeps(className)
+            .flatMap: libClassName =>
+              relations.libraryClassName
+                .reverse(libClassName)
+                .map(stampVf)
+          val classDigests = relations.productClassName
+            .reverse(className)
+            .flatMap: prodClassName =>
+              relations
+                .definesClass(prodClassName)
+                .flatMap: sourceFile =>
+                  relations
+                    .products(sourceFile)
+                    .map(stampVf)
```

Comment:
So we use the hash of the source file to stamp the class? This means we will have slightly more invalidation than needed. Don't we have the hash of the class file anywhere?

Reply:
Zinc might know, but I don't think we're exposing it? See the commented out code below.

Reply:
Okay